cmd/root/options.go

```diff
@@ -0,0 +1,134 @@
+package root
+
+import (
+	"errors"
+	"fmt"
+	"net/http"
+	"net/http/pprof"
+	"os"
+	"path/filepath"
+	"reflect"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/serializer"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/webhook"
+
+	configv1alpha1 "github.com/grafana/tempo-operator/api/config/v1alpha1"
+)
+
+var errConfigFileLoading = errors.New("could not read file at path")
+
+func loadConfigFile(scheme *runtime.Scheme, outConfig *configv1alpha1.ProjectConfig, configFile string) error {
+	content, err := os.ReadFile(filepath.Clean(configFile))
+	if err != nil {
+		return fmt.Errorf("%w %s", errConfigFileLoading, configFile)
+	}
+
+	codecs := serializer.NewCodecFactory(scheme)
+
+	if err = runtime.DecodeInto(codecs.UniversalDecoder(), content, outConfig); err != nil {
+		return fmt.Errorf("could not decode file into runtime.Object: %w", err)
+	}
+
+	return nil
+}
+
+// LoadConfig initializes the controller configuration, optionally overriding the defaults
+// from a provided configuration file.
+func LoadConfig(scheme *runtime.Scheme, configFile string) (*configv1alpha1.ProjectConfig, ctrl.Options, error) {
+	options := ctrl.Options{Scheme: scheme}
+	ctrlConfig := configv1alpha1.DefaultProjectConfig()
+
+	if configFile == "" {
+		return &ctrlConfig, options, nil
+	}
+
+	err := loadConfigFile(scheme, &ctrlConfig, configFile)
+	if err != nil {
+		return nil, options, fmt.Errorf("failed to parse controller manager config file: %w", err)
+	}
+
+	err = ctrlConfig.Validate()
+	if err != nil {
+		return nil, options, fmt.Errorf("controller config validation failed: %w", err)
+	}
+
+	options = mergeOptionsFromFile(options, &ctrlConfig)
+	return &ctrlConfig, options, nil
+}
+
+func mergeOptionsFromFile(o manager.Options, cfg *configv1alpha1.ProjectConfig) manager.Options {
+	o = setLeaderElectionConfig(o, cfg.ControllerManagerConfigurationSpec)
+
+	if o.Metrics.BindAddress == "" && cfg.Metrics.BindAddress != "" {
+		o.Metrics.BindAddress = cfg.Metrics.BindAddress
+
+		endpoints := map[string]http.HandlerFunc{
+			"/debug/pprof/":        pprof.Index,
+			"/debug/pprof/cmdline": pprof.Cmdline,
+			"/debug/pprof/profile": pprof.Profile,
+			"/debug/pprof/symbol":  pprof.Symbol,
+			"/debug/pprof/trace":   pprof.Trace,
+		}
+
+		if o.Metrics.ExtraHandlers == nil {
+			o.Metrics.ExtraHandlers = map[string]http.Handler{}
+		}
+
+		for path, handler := range endpoints {
+			o.Metrics.ExtraHandlers[path] = handler
```

Comment:
I'm not sure if I fully understand this part, does this always enable the pprof endpoint if operator metrics are enabled?

If so, why? imho pprof should be enabled explicitly, not always when metrics are enabled.

Reply:
that is actually a good point. I'll review this part and see what is the appropiate behavriour, I think we can add a flag ,

Reply:
Just removed the pprof endpoint for now.